### DISCUSSION TITLE
🔒 Security: becoming independent from bcrypt package

### DISCUSSION THREAD

**Discussion by jankapunkt** [meteor/meteor#13529] [Feature Requests]
Recently I discovered, that `bcrypt` is a cause of some warnings with the latest Meteor 3.0.4 and Node 22. When installing npm dependencies I got:

```shell
$ meteor npm install
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
```

So I checked `npm list inflight`:

```shell
└─┬ bcrypt@5.1.1
  └─┬ @mapbox/node-pre-gyp@1.0.11
    └─┬ rimraf@3.0.2
      └─┬ glob@7.2.3
        └── inflight@1.0.6
```

So I discovered that `bcrypt` is still [not updated since a year](https://www.npmjs.com/package/bcrypt) (uh oh!) and there are some issues that raise my concern :

- https://github.com/kelektiv/node.bcrypt.js/issues/1053
- https://github.com/kelektiv/node.bcrypt.js/issues/1058
- https://github.com/kelektiv/node.bcrypt.js/pulls (many of them not answered or handled by the author)

It seems that it is on its way to abandonware and **this can become a big problem for all of us if we don't talk about this now!**

My questions to the community but also to Meteor core team are:

- how can we prepare for `bcrypt` becoming abandondened or even vulnarable?
- do we want to keep accounts being tightly coupled to bcrypt or make it a default, becoming able to to use alternatives, such as [argon2](https://www.npmjs.com/package/argon2)

**Comment by xet7** [comment]
@jankapunkt 

For bcrypt, does Javascript version have same problems?

- https://www.npmjs.com/package/bcryptjs
- https://www.npmjs.com/package/@types/bcryptjs

**Reply by jankapunkt** [reply]
Yes it's even worse, 8 years last release

**Comment by StorytellerCZ** [comment]
Not a fire right now, but it is concerning. Given that this is the core when it comes to account passwords and in general in the JS ecosystem.
No matter what we should be looking for alternatives. Maybe we are missing that there has been recommendation to migrate to something else, or there might be better alternatives.

What about `node:crypto`?

**Reply by jesusej** [reply]
I've done a quick research and found this [OWASP article about password storage](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html). It dives a bit more about password hashing but at the end they recommend first argon2, then scrypt (node:crypto) and at the end bcrypt for legacy systems

**Reply by StorytellerCZ** [reply]
In this regard we are the legacy system here. Argon2 is worth investigation.

**Reply by StorytellerCZ** [reply]
Here is a migration to Argon2: https://github.com/ranisalt/node-argon2/wiki/Migrating-from-another-hash-function

**Reply by jankapunkt** [reply]
I think this all seems to be a valid candidate for a strategy pattern implementation in accounts-password

**Comment by vparpoil** [comment]
I just made a PR to use `argon2` to store passwords :
https://github.com/meteor/meteor/pull/13554

**Reply by jankapunkt** [reply]
Is it backwards compatible with existing accounts?

**Reply by vparpoil** [reply]
Yes of course, the existing passwords get migrated as the users log in. 
There is no way to migrate all data because obviously we don’t know the passwords

**Reply by StorytellerCZ** [reply]
We will have to make sure that everyone knows that they need to get their users to sign-in. It would be a nice to point out that people can write emails telling their users on how they take security seriously, that here is no breach, but they need to make sure that they logout and login for the transition to happen.

**Reply by vparpoil** [reply]
@StorytellerCZ I could add a note to the PR. In which readme file would that be appropriate ? I think a Changelog note would be appropriate, but I don't see where to add this note to be taken into account in the changelog

**Reply by nachocodoner** [reply]
@vparpoil: We’ve discussed the change internally, and our concern is that it’s not only a breaking change in code but also in data, which could have a bigger impact on production environments and reaction times.

To address this, we could introduce it as an opt-in feature initially, allowing developers to adapt and manage the incremental migration risks, and any possible side-effect we may not count now. It may be better if bcrypt management isn’t completely removed in the PR, but instead offers a configurable option to choose one method, with bcrypt as the default.

If this becomes a critical change, we may see the user base favoring argon2, and we could plan a more structured transition away from bcrypt at that point.

What are your thoughts on this?

**Reply by vparpoil** [reply]
We could add a feature toggle, but I enabling it by default at some point is crucial.

What do you think about the following :
- start with a opt-in feature toggle like Accounts.config({argon2Enabled: true})
- provide clear docs about the migration process, including monitoring tips
- set a timeline when the feature toggle becomes true by default (of course while collecting feedback)

**Reply by nachocodoner** [reply]
@vparpoil: Sounds great! Lets do like this

**Reply by vparpoil** [reply]
Ok, Thanks, will work on updating the PR on the next days

**Reply by jankapunkt** [reply]
I'm ready to test this with our staging env, once there is a stable Beta/RC release out.

**Reply by paulincai** [reply]
I can test this in production and migrate permanently to it at any time.

**Reply by StorytellerCZ** [reply]
I will need a longer migration window for my production apps.

**Reply by vparpoil** [reply]
I just pushed some more commits to : 
- add the feature toggle
- revert the changes in tests to the initial version
- add a new test file dedicated to tests related to passwords with argon
- add more documentation on this feature toggle and the migration process
- added the ability to rollback from argon to bcrypt if needed

**Reply by jankapunkt** [reply]
If someone from meteor core team releases a alpha version of the package I would start testing in real world conditions

**Reply by nachocodoner** [reply]
[I released `3.2-beta.1`](https://github.com/meteor/meteor/pull/13610) (`meteor update --release 3.2-beta.1`) including this feature.

You can already test this feature. I will do some testing in coming weeks.

**Reply by nachocodoner** [reply]
This change is available in Meteor 3.2.

To use it, run: `meteor update --release 3.2` or `meteor create --release 3.2`.

Find instructions on how to enable it in the [updated docs 6](https://github.com/meteor/meteor/blob/9331c224af0b590f96fa0fa654e724d2fb3dc41c/v3-docs/docs/api/accounts.md#passwords-passwords).

Continue the discussion here for further support and changes.

**Reply by afaqahmedkhan** [reply]
> This change is available in Meteor 3.2.
> 
> To use it, run: `meteor update --release 3.2` or `meteor create --release 3.2`.
> 
> Find instructions on how to enable it in the [updated docs 6](https://github.com/meteor/meteor/blob/9331c224af0b590f96fa0fa654e724d2fb3dc41c/v3-docs/docs/api/accounts.md#passwords-passwords).
> 
> Continue the discussion here for further support and changes.

Hey @nachocodoner , thanks for this update. after I update to 3.2-beta.1 and versions updated too. I try enabling argon2Enabled as you mentioned in the docs. unfortunately it gives me an error. _Object literal may only specify known properties, and 'argon2Enabled' does not exist in type_
is there something I am missing? thanks for the help and support 
<img width="988" alt="image" src="https://github.com/user-attachments/assets/905b9230-f3c5-47a6-baf0-5bc1249a1553" />
<img width="736" alt="image" src="https://github.com/user-attachments/assets/afea9bbe-3484-42ad-acfc-842d19bb4425" />

**Reply by vparpoil** [reply]
Your accounts-password is not on the right version I think. Try to update to 3.2 final and not the beta
Here is what I have with meteor 3.2
> accounts-base@3.1.0
> accounts-password@3.1.0

**Reply by afaqahmedkhan** [reply]
Hi @vparpoil thank you for your response. It works for me

**Comment by StorytellerCZ** [comment]
Looks like bcrypt v6 will be released soon: https://github.com/kelektiv/node.bcrypt.js/releases/tag/v6.0.0
